daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala

```diff
@@ -298,231 +299,217 @@ class Runner(
         Set.empty,
       )
 
-    // Removing the early return only makes this harder to read.
-    @SuppressWarnings(Array("org.wartremover.warts.Return"))
-    def stepToValue(): Either[RuntimeException, Unit] = {
-      while (!machine.isFinal) {
+    @tailrec
+    def stepToValue(): Future[SValue] =
+      if (machine.isFinal())
```

Comment:
This code looks weird, why do we need the recursive call? Can’t I just call `machine.run()` at the beginning which will immediately return if I have a final value and never recurse?

Reply:
@cocreature is right, we don't need to call `machine.isFinal` anymore. `machine.run` will run until we have a value or need some input. The code here looks like we never handle the need for more input anyway. Thus, calling `machine.run` once should do the job.

Reply:
I was confuse. I fix that right now.